File: ClassEval_T/java/solutuon/ShoppingCart.java
package org.example;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public class ShoppingCart {
    private Map<String, Item> items;

    public ShoppingCart() {
        this.items = new HashMap<>();
    }

    public void addItem(String item, double price, int quantity) {
        if (items.containsKey(item)) {
            Item existingItem = items.get(item);
            existingItem.setPrice(price);
            existingItem.setQuantity(existingItem.getQuantity() + quantity);
        } else {
            items.put(item, new Item(price, quantity));
        }
    }

    public void removeItem(String item, int quantity) {
        if (items.containsKey(item)) {
            Item existingItem = items.get(item);
            int newQuantity = existingItem.getQuantity() - quantity;
            if (newQuantity <= 0) {
                items.remove(item);
            } else {
                existingItem.setQuantity(newQuantity);
            }
        }
    }

    public Map<String, Item> viewItems() {
        return new HashMap<>(items);
    }

    public double totalPrice() {
        double total = 0.0;
        for (Item item : items.values()) {
            total += item.getPrice() * item.getQuantity();
        }
        return total;
    }

    public static class Item {
        private double price;
        private int quantity;

        public Item(double price, int quantity) {
            this.price = price;
            this.quantity = quantity;
        }

        public double getPrice() {
            return price;
        }

        public void setPrice(double price) {
            this.price = price;
        }

        public int getQuantity() {
            return quantity;
        }

        public void setQuantity(int quantity) {
            this.quantity = quantity;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Item item = (Item) o;
            return Double.compare(item.price, price) == 0 && quantity == item.quantity;
        }

        @Override
        public int hashCode() {
            return Objects.hash(price, quantity);
        }
    }
}

File: ClassEval_T/cpp/solution/ShoppingCart.cpp
#include <unordered_map>
#include <string>

class ShoppingCart {
public:
    ShoppingCart();

    void add_item(const std::string& item, double price, int quantity = 1);
    void remove_item(const std::string& item, int quantity = 1);
    std::unordered_map<std::string, std::pair<double, int>> view_items() const;
    double total_price() const;

private:
    std::unordered_map<std::string, std::pair<double, int>> items;
};

ShoppingCart::ShoppingCart() {}

void ShoppingCart::add_item(const std::string& item, double price, int quantity) {
    if (items.find(item) != items.end()) {
        items[item] = { price, quantity };
    }
    else {
        items[item] = { price, quantity };
    }
}

void ShoppingCart::remove_item(const std::string& item, int quantity) {
    auto it = items.find(item);
    if (it != items.end()) {
        it->second.second -= quantity;
        if (it->second.second <= 0) {
            items.erase(it);
        }
    }
}

std::unordered_map<std::string, std::pair<double, int>> ShoppingCart::view_items() const {
    return items;
}

double ShoppingCart::total_price() const {
    double total = 0.0;
    for (const auto& [item, info] : items) {
        total += info.first * info.second;
    }
    return total;
}
